### PR TITLE
fix #13150 Do not add schema / class name mapping where custom mapping exists

### LINKED ISSUE #13150: Spring generates duplicated JsonSubTypes
Hello everyone,

When I use the openapi-generator with a discriminator and mapping, it generates two @JsonSubTypes

```
discriminator:
    propertyName: type
    mapping:
         foo: "#components/schemas/Foo"
```

Generates something like this

```
@JsonSubTypes.Type(value = Foo.class, name = "Foo")
@JsonSubTypes.Type(value = Foo.class, name = "foo")
```

Notice the duplication. The Foo is the class name of the class, and the lowercase foo is my mapping. Wtf is going on lol. This doesn't seem right.

The `Foo` class is defined something like

```
Foo:
    allOf:
         - $ref: "#components/schemas/someParent"
         - type: object
            properties:
                type:
                    type: string
```

I am using version 6.0.1 and the schema version is `openapi: '3.0.0'`

Here's the full spec to generate the incorrect thing:
```
openapi: '3.0.0'
info:
  version: '1.0.0'
  title: 'FooService'
paths:
  /parent:
    put:
      summary: put parent
      operationId: putParent
      parameters:
        - name: name
          in: path
          required: true
          description: Name of the account being updated.
          schema:
            type: string
      requestBody:
        description: The updated account definition to save.
        required: true
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/Parent'
      responses:
        '200':
          $ref: '#/components/responses/Parent'
components:
  schemas:
    Parent:
      type: object
      description: Defines an account by name.
      properties:
        name:
          type: string
          description: The account name.
        type:
          type: string
          description: The account type discriminator.
      required:
        - name
        - type
      discriminator:
        propertyName: type
        mapping:
          foo: '#/components/schemas/Foo'

    Foo:
      allOf:
        - $ref: "#/components/schemas/Parent"
        - type: object
          properties:
            fooType:
              type: string 
  responses:
    Parent:
      description: The saved account definition.
      content:
        application/json:
          schema:
            $ref: '#/components/schemas/Parent'
```

I generated this by running
`java -jar openapi-generator-cli.jar generate -i schema.yaml -o foo -g spring`

### PR DESCRIPTION
**Pull request by bernie-schelberg-mywave**
Having multiple names for a model mapping can cause issues with clients that are expecting a particular types to always have the same model name. Refer https://github.com/OpenAPITools/openapi-generator/issues/13150.

### PR checklist
 
- [X] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [X] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [X] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [X] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.1.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [X] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by bernie-schelberg-mywave** [comment]
@cachescrubber (2022/02) @welshm (2022/02) @MelleD (2022/02) @atextor (2022/02) @manedev79 (2022/02) @javisst (2022/02) @borsch (2022/02) @banlevente (2022/02) @Zomzog (2022/09)

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java
Comment (welshm) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:3370]: This logic is changing - since this value previously would have been `null`

**Review by welshm — APPROVED** [review]

**Comment by welshm** [comment]
Is this still required after https://github.com/OpenAPITools/openapi-generator/pull/13958 ?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java
Comment (bernie-schelberg-mywave) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:3370]: No, prior to this change `currentSchemaName` was added as a mapping name on line 3365, and then another mapping was added on line 3371 if an `x-discriminator-value` was set. With this change, it's one or the other, not both.

**Comment by bernie-schelberg-mywave** [comment]
> Is this still required after #13958 ?

They seem unrelated to me. The other is about adding additional annotations to the model class / interface. This one is about having singular mapping names for `oneOf` subclasses.

**Comment by wing328** [comment]
cc @OpenAPITools/generator-core-team as the change impacts default codegen.

**Comment by jorgerod** [comment]
Hello @bernie-schelberg-mywave @wing328 

What is the status of this issue? 
IMHO this is a bug and should be fixed as soon as possible. We have been blocked by this issue since 5.3.1 and I think this problem will happen to many teams.

**Comment by bernie-schelberg-mywave** [comment]
@jorgerod as far as I know, it is planned to merge this one into the version 7 branch. I don't really have any update on a timeline though, sorry.

**Comment by jorgerod** [comment]
Hi @bernie-schelberg-mywave 

I have done a serialization test with your changes. The JsonSubTypes are no longer duplicated.
I used the spec from this PR and the test is as follows.

```java
  @Test
  void testSerialization() throws JsonProcessingException {
    final ObjectMapper objectMapper = new ObjectMapper();

    final String foo = objectMapper.writeValueAsString(new FooDTO());
    assertThat(foo).contains("{\"type\":\"Foo\"}");

  }
```

The output is:

```
java.lang.AssertionError: 
Expecting actual:
  "{"type":"Foo"}"
to contain:
  "{"type":"foo"}" 
```

I'm not a Jackson expert, but it seems that in the serialization Jackson is using the value of `@JsonTypeName` instead of the value defined in the mapping.

What do you think?

**Comment by bernie-schelberg-mywave** [comment]
Hi @jorgerod, the error message in your comment doesn't match your test. In your test you say that you expect `foo` to contain `{"type":"Foo"}`, but the assertion error states that it's expecting `{"type":"foo"}`. The `actual` value in the assertion error matches the assertion that you have in the test. Can you check and clarify what it is that you're expecting?

Also, what is `FooDTO`? It's not specified in the schema that is part of this PR. Have you defined it separately?

**Comment by troehling** [comment]
Hi @bernie-schelberg-mywave, I have experienced a similar issue as @jorgerod. Regarding your question of the `FooDTO`: In my case I have configured the suffix `dto` as `modelNameSuffix` in my generator. This leads to `@JsonTypeName("Foo")` being generated, which is used by Jackson for the serialization overriding the value configured in `@JsonSubTypes`.

**Comment by jorgerod** [comment]
> Hi @bernie-schelberg-mywave
> 
> I have done a serialization test with your changes. The JsonSubTypes are no longer duplicated. I used the spec from this PR and the test is as follows.
> 
> ```java
>   @Test
>   void testSerialization() throws JsonProcessingException {
>     final ObjectMapper objectMapper = new ObjectMapper();
> 
>     final String foo = objectMapper.writeValueAsString(new FooDTO());
>     assertThat(foo).contains("{\"type\":\"Foo\"}");
> 
>   }
> ```
> 
> The output is:
> 
> ```
> java.lang.AssertionError: 
> Expecting actual:
>   "{"type":"Foo"}"
> to contain:
>   "{"type":"foo"}" 
> ```
> 
> I'm not a Jackson expert, but it seems that in the serialization Jackson is using the value of `@JsonTypeName` instead of the value defined in the mapping.
> 
> What do you think?

Hi @bernie-schelberg-mywave 
In the example, in the hurry, there is a typo.

spec
```yaml
openapi: '3.0.0'
info:
  version: '1.0.0'
  title: 'FooService'
paths:
  /parent:
    put:
      summary: put parent
      operationId: putParent
      parameters:
        - name: name
          in: path
          required: true
          description: Name of the account being updated.
          schema:
            type: string
      requestBody:
        description: The updated account definition to save.
        required: true
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/Parent'
      responses:
        '200':
          $ref: '#/components/responses/Parent'
components:
  schemas:
    Parent:
      type: object
      description: Defines an account by name.
      properties:
        name:
          type: string
          description: The account name.
        type:
          type: string
          description: The account type discriminator.
      required:
        - name
        - type
      discriminator:
        propertyName: type
        mapping:
          foo: '#/components/schemas/Foo'

    Foo:
      allOf:
        - $ref: "#/components/schemas/Parent"
        - type: object
          properties:
            fooType:
              type: string
  responses:
    Parent:
      description: The saved account definition.
      content:
        application/json:
          schema:
            $ref: '#/components/schemas/Parent'
```

Test:
```java
  @Test
  void testSerialization() throws JsonProcessingException {
    final ObjectMapper objectMapper = new ObjectMapper();

    final String foo = objectMapper.writeValueAsString(new FooDTO());
    assertThat(foo).contains("{\"type\":\"foo\"}");

  }
```

Output:
```
java.lang.AssertionError: 
Expecting actual:
  "{"type":"foo"}"
to contain:
  "{"type":"Foo"}" 
```

As @troehling says, Jackson use @JsonTypeName("Foo") for serialization.

I'm testing and I think that modifying `pojo.mustache` for the discriminator case with mapping will solve it.

Currently there is this:
https://github.com/OpenAPITools/openapi-generator/blob/12e76ec14f827cc126acca22203eb4d382402e7d/modules/openapi-generator/src/main/resources/JavaSpring/pojo.mustache#L16-L20

It would be to put this:
```mustache
{{#jackson}}
{{#isClassnameSanitized}}
    {{^hasDiscriminatorWithNonEmptyMapping}}
@JsonTypeName("{{name}}")
    {{/hasDiscriminatorWithNonEmptyMapping}}
{{/isClassnameSanitized}}
{{/jackson}}
```
If there is a discriminator with mapping, don't add @JsonTypeInfo

I also think this bugfix should go in a separate PR from this one as they are different problems.

How do you see it @bernie-schelberg-mywave @troehling @wing328 @welshm ?

Thank you

**Comment by bernie-schelberg-mywave** [comment]
@jorgerod I think I understand now. I think it may be related, but a different issue to this one. I think the issue you describe is present in the master branch, is it not?

**Comment by welshm** [comment]
Sorry this fell off my radar for a bit! Can you regenerate the samples to include with the PR? Or did no samples change? This looks fine to me from a SpringCodegen point of view - but the changes are in the default codegen so may need more reviewers - although @wing328 may be able to take a look as well

**Comment by welshm** [comment]
@jorgerod I agree with your suggestion that the fix should be in a separate PR - it will also be quicker to merge and simpler to review as it will just be touching a Spring codegen mustache.

**Comment by bernie-schelberg-mywave** [comment]
@welshm I merged `master` again, and re-generated samples. There were no changes.

**Comment by welshm** [comment]
> @welshm I merged `master` again, and re-generated samples. There were no changes.

Might want to add some samples then that show this behavior being used. Not required though.

**Comment by bernie-schelberg-mywave** [comment]
@welshm I modified one of the examples to have a discriminator mapping, renaming `className` to `type` in the process. That changed more than was really necessary, but `className` didn't seem right when the class name was no longer being used as the discriminator.
Generated all samples and pushed the changes. I noticed that the cs generator seems not to use the discriminator mapping values, but that's a separate issue I think.

**Comment by jorgerod** [comment]
Hi @bernie-schelberg-mywave @welshm 
I have created this [PR ](https://github.com/OpenAPITools/openapi-generator/pull/14733)

File: modules/openapi-generator/src/test/resources/3_0/petstore-with-fake-endpoints-models-for-testing.yaml
Comment (welshm) [inline modules/openapi-generator/src/test/resources/3_0/petstore-with-fake-endpoints-models-for-testing.yaml:1493]: nit: I would not use `type` here since it's the name of another property of the YAML - maybe `animalType` ?

File: samples/client/petstore/java-helidon-client/mp/src/main/java/org/openapitools/client/model/Cat.java
Comment (welshm) [inline samples/client/petstore/java-helidon-client/mp/src/main/java/org/openapitools/client/model/Cat.java:24]: Seems weird to import itself, but separate problem

File: samples/client/petstore/java/apache-httpclient/api/openapi.yaml
Comment (welshm) [inline samples/client/petstore/java/apache-httpclient/api/openapi.yaml:1515]: Same nit - would probably call this `animalType` or `animalSpecies`

**Review by welshm — COMMENTED** [review]
Thank you for updating the samples (and modifying them to use excercise the code change!)

Minor suggestion on parameter naming, but otherwise I think this is good.

If you are _not_ on the Slack, I would recommend joining - [Link](https://join.slack.com/t/openapi-generator/shared_invite/zt-1poli1esx-wDJf2MEKs8bjpd~t1OwVIg) 

You will need to get someone from the default codegen maintainers to review still

File: modules/openapi-generator/src/test/resources/3_0/petstore-with-fake-endpoints-models-for-testing.yaml
Comment (bernie-schelberg-mywave) [reply modules/openapi-generator/src/test/resources/3_0/petstore-with-fake-endpoints-models-for-testing.yaml]: agree, it's confusing. I've changed it to `species`.

File: samples/client/petstore/java-helidon-client/mp/src/main/java/org/openapitools/client/model/Cat.java
Comment (bernie-schelberg-mywave) [reply samples/client/petstore/java-helidon-client/mp/src/main/java/org/openapitools/client/model/Cat.java:24]: Yeah, that's weird. I've seen that some subclasses have the `@JsonSubTypes` annotation as well (samples/client/petstore/java/feign/src/main/java/org/openapitools/client/model/Cat.java for example), which I don't think is necessary (not because of this change, the `master` branch shows this behaviour as well, but it wasn't in the samples before because the source schema didn't use discriminator values). Because of this annotation, every subclass refers to every other subclass. This may be the reason for the unnecessary import here.

File: samples/client/petstore/java/apache-httpclient/api/openapi.yaml
Comment (bernie-schelberg-mywave) [reply samples/client/petstore/java/apache-httpclient/api/openapi.yaml]: changed to `species`

**Comment by bernie-schelberg-mywave** [comment]
@OpenAPITools/generator-core-team please review.

@wing328 , @welshm , I'm not sure I'm using the right handle, how do I cc / request review from the core team?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java
Comment (welshm) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:3441]: I wonder if this change is causing the issues you're seeing with Dart. I'd have to look more closely at how the schemas are being used later on, but this adds _less_ schemas now; so perhaps some filterins logic is applied at some point?

I would try reverting this part of the change and see what the Dart output looks like.

**Comment by wing328** [comment]
I've updated the target branch to 7.0.x as I think that's what this change is based on.

File: samples/client/petstore/java-helidon-client/mp/src/main/java/org/openapitools/client/model/Cat.java
Comment (wing328) [reply samples/client/petstore/java-helidon-client/mp/src/main/java/org/openapitools/client/model/Cat.java:24]: agreed it's something we can deal with later.

File: samples/client/petstore/java/apache-httpclient/api/openapi.yaml
Comment (wing328) [reply samples/client/petstore/java/apache-httpclient/api/openapi.yaml]: i've cherry-pick some commits and created https://github.com/OpenAPITools/openapi-generator/pull/14984 instead.

i didn't rename `className` to something else

**Comment by wing328** [comment]
closed via https://github.com/OpenAPITools/openapi-generator/pull/14984 instead

**Pull request closed without merging**
